safeweb/http.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package safeweb provides a simple, safe, and opinionated wrapper around
+// http.Server that applies some web application security defenses by default.
+// It is intended to be used in place of http.Server for internal applications
+// that reside on the Tailscale network.
+//
+// safeweb requires that applications be built using two distinct
+// http.ServeMuxes for serving browser and API clients to serve the appropriate
+// security defenses in their respective contexts. When serving requests it
+// will first attempt to route requests using the browser mux, and if no
+// matching route is found it will instead defer to the API mux.
+//
+// safeweb enforces Cross-Site Request Forgery (CSRF) protection for all routes
+// in the browser mux using the gorilla/csrf package. It is necessary to
+// template the CSRF token into all forms that are submitted to the server using
+// the gorilla/csrf `csrf.TemplateField` and `csrf.TemplateTag(r *http.Request)`
+// APIs. Applications that are served over a secure context (i.e. HTTPS) should
+// specify the SecureContext field in the configuration to ensure that the
+// Secure flag is set on CSRF cookies.
+//
+// safeweb will apply the following to requests served by the browser mux:
+//   - A Content-Security-Policy header that disallows inline scripts, framing, and third party resources.
+//   - CSRF protection for all forms.
+//   - X-Content-Type-Options header on responses set to "nosniff" to prevent MIME type sniffing attacks.
+//
+// safeweb will apply the following to requests served by the API mux:
+//   - Content-Type header validation to prohibit `application/x-www-form-urlencoded` requests.
+//   - Cross-Origin Resource Sharing headers on responses if they are provided in the configuration.
+//
+// example usage:
+//
+//	h := http.NewServeMux()
+//	h.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+//		fmt.Fprint(w, "Hello, world!")
+//	})
+//	s, err := safeweb.NewServer(safeweb.Config{
+//		BrowserMux: h,
+//	})
+//	if err != nil {
+//		log.Fatalf("failed to create server: %v", err)
+//	}
+//	ln, err := net.Listen("tcp", ":8080")
+//	if err != nil {
+//		log.Fatalf("failed to listen: %v", err)
+//	}
+//	defer ln.Close()
+//	if err := s.Serve(ln); err != nil && err != http.ErrServerClosed {
+//		log.Fatalf("failed to serve: %v", err)
+//	}
+package safeweb
+
+import (
+	crand "crypto/rand"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gorilla/csrf"
+)
+
+// the default Content-Security-Policy header.
```

Comment:
It's unexported so I don't insist on "doc style" but please do capitalize 😆 
```suggestion
// The default Content-Security-Policy header.
```

(this is 100% an optional aesthetic nit)